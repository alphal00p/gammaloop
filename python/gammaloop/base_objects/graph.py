from __future__ import annotations

import gammaloop.base_objects.model as model
from gammaloop.misc.common import *

# Abstract class for all vertex info
class VertexInfo(object):
    pass

class InteractonVertexInfo(VertexInfo):
    def __init__(self, vertex_rule: model.VertexRule):
        self.vertex_rule: model.VertexRule = vertex_rule

class Vertex(object):
    def __init__(self, name: str, vertex_info: VertexInfo, edges: list[Edge] | None):
        self.name: name = name
        self.vertex_rule: VertexInfo = vertex_info
        self.edges: list[Edge] = edges

class Edge(object):
    def __init__(self, name: str, particle: model.Particle, vertices: (Vertex, Vertex) | None):
        self.id: name = name
        self.particle: model.Particle = particle
        self.vertices: (Vertex, Vertex) = vertices

# TODO create serializable counterpart and export to yaml, similarly as done for the model.

class Graph(object):
    def __init__(self, name: str, vertices: list[Vertex], edges: list[Edge]):
        self.name: str = name
        self.vertices: list[Vertex] = vertices
        self.edges: list[Edge] = edges
        self.name_to_position: dict[str,dict[str,int]] = {}
    
    def get_edge(self, edge_name: str) -> Edge:
        return self.edges[self.name_to_position['edges'][edge_name]]
    
    def get_vertex(self, vertex_name: str) -> Parameter:
        return self.vertices[self.name_to_position['vertices'][vertex_name]]

    def synchronize_name_map(self) -> None:
        for i_e, edge in enumerate(self.edges):
            self.name_to_position['edges'][edge.name] = i_e
        for i_v, vertex in enumerate(self.vertices):
            self.name_to_position['vertices'][vertex.name] = i_v


    @staticmethod
    def from_qgraph(model: model.Model, qgraph_object) -> Graph:
        """ Imports graph form a stylicized qgraph python output file. Will be deprecated when using in-house gammaloop graph generation. """
        logger.info("Importing graph from qgraph python output")
        # TODO
        return Graph('default',[],[])