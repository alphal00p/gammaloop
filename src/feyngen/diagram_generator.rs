use log::debug;

use super::{FeynGenError, FeynGenOptions};
use crate::{graph::BareGraph, model::Model};
use symbolica::graph::Graph as SymbolicaGraph;

pub struct FeynGen {
    pub options: FeynGenOptions,
}

impl FeynGen {
    pub fn new(options: FeynGenOptions) -> Self {
        Self { options }
    }

    pub fn generate(&self, model: &Model) -> Result<Vec<BareGraph>, FeynGenError> {
        debug!(
            "Generating Feynman diagrams for model {} and process:\n{}",
            model.name, self.options
        );

        let vertex_signatures = model
            .vertex_rules
            .iter()
            .map(|v| {
                v.particles
                    .iter()
                    .map(|p| p.name.as_str())
                    .collect::<Vec<_>>()
            })
            .collect::<Vec<_>>();

        let mut external_edges = self
            .options
            .initial_pdgs
            .iter()
            .enumerate()
            .map(|(i_initial, pdg)| {
                (
                    i_initial + 1,
                    model
                        .get_particle_from_pdg(*pdg as isize)
                        .clone()
                        .name
                        .clone(),
                )
            })
            .collect::<Vec<_>>();
        external_edges.extend(
            self.options
                .final_pdgs
                .iter()
                .enumerate()
                .map(|(i_final, pdg)| {
                    (
                        self.options.initial_pdgs.len() + i_final + 1,
                        model
                            .get_particle_from_pdg(*pdg as isize)
                            .clone()
                            .name
                            .clone(),
                    )
                })
                .collect::<Vec<_>>(),
        );
        debug!("external_edges = {:?}", external_edges);
        debug!("vertex_signatures = {:?}", vertex_signatures);

        let graphs = SymbolicaGraph::generate(
            &external_edges
                .iter()
                .map(|(i, name)| (*i, name.as_str()))
                .collect::<Vec<_>>(),
            &vertex_signatures,
            None,
            Some(self.options.loop_count_range.1),
            self.options.filters.get_max_bridge(),
            self.options.filters.allow_tadpoles(),
        );

        debug!("Symbolica generated {} graphs", graphs.len());

        Ok(vec![])
    }
}
