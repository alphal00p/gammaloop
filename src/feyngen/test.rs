use std::sync::Arc;

use ahash::HashMap;
use ahash::HashMapExt;
use symbolica::graph::Graph;

use crate::feyngen::diagram_generator::EdgeColor;
use crate::model::ColorStructure;
use crate::model::VertexRule;
use crate::tests_from_pytest::load_generic_model;

use super::diagram_generator::NodeColorWithVertexRule;
use super::GenerationType;
use super::{diagram_generator::FeynGen, FeynGenFilter, FeynGenFilters, FeynGenOptions};

#[test]
fn cut_content() {
    let model = load_generic_model("sm");

    let mut coupling = HashMap::new();
    coupling.insert("QED".into(), 6);
    let mut pert = HashMap::new();
    pert.insert("QCD".into(), 1);
    let filters = FeynGen::new(FeynGenOptions {
        generation_type: GenerationType::CrossSection,
        initial_pdgs: vec![-11, 11],
        final_pdgs: vec![5, -5, 25],
        loop_count_range: (3, 3),
        symmetrize_initial_states: true,
        symmetrize_final_states: true,
        symmetrize_left_right_states: true,
        allow_symmetrization_of_external_fermions_in_amplitudes: false,
        max_multiplicity_for_fast_cut_filter: 6,
        amplitude_filters: FeynGenFilters(vec![
            FeynGenFilter::SelfEnergyFilter(Default::default()),
            FeynGenFilter::ParticleVeto(vec![
                23, 24, 9000001, 9000002, 9000003, 9000004, 12, 14, 16, 2, 4, 6, 3, 250, 251, 13,
                15,
            ]),
            FeynGenFilter::TadpolesFilter(Default::default()),
            FeynGenFilter::ZeroSnailsFilter(Default::default()),
        ]),
        cross_section_filters: FeynGenFilters(vec![
            FeynGenFilter::SelfEnergyFilter(Default::default()),
            FeynGenFilter::ParticleVeto(vec![
                23, 24, 9000001, 9000002, 9000003, 9000004, 12, 14, 16, 2, 4, 6, 3, 250, 251, 13,
                15,
            ]),
            FeynGenFilter::TadpolesFilter(Default::default()),
            FeynGenFilter::ZeroSnailsFilter(Default::default()),
            FeynGenFilter::PerturbativeOrders(pert),
            FeynGenFilter::CouplingOrders(coupling),
            FeynGenFilter::LoopCountRange((3, 3)),
        ]),
    });

    let mut graph = Graph::new();
    #[allow(non_snake_case)]
    let bbH = NodeColorWithVertexRule {
        external_tag: 0,
        vertex_rule: model.get_vertex_rule(&"V_78".into()),
    };
    let bba = NodeColorWithVertexRule {
        external_tag: 0,
        vertex_rule: model.get_vertex_rule(&"V_73".into()),
    };
    let bbg = NodeColorWithVertexRule {
        external_tag: 0,
        vertex_rule: model.get_vertex_rule(&"V_76".into()),
    };
    let epema = NodeColorWithVertexRule {
        external_tag: 0,
        vertex_rule: model.get_vertex_rule(&"V_98".into()),
    };

    let dummy_external_vertex_rule = Arc::new(VertexRule {
        name: "external".into(),
        couplings: vec![],
        lorentz_structures: vec![],
        particles: vec![],
        color_structures: ColorStructure::new(vec![]),
    });

    let v0 = graph.add_node(bbH.clone());
    let v1 = graph.add_node(bbH.clone());
    let v2 = graph.add_node(bbg.clone());
    let v3 = graph.add_node(bba.clone());
    let v4 = graph.add_node(bbg.clone());
    let v5 = graph.add_node(bba.clone());
    let v6 = graph.add_node(epema.clone());
    let v7 = graph.add_node(epema.clone());
    let e1 = NodeColorWithVertexRule {
        external_tag: 1,
        vertex_rule: dummy_external_vertex_rule.clone(),
    };
    let mut e2 = e1.clone();
    e2.external_tag = 2;
    let mut e3 = e1.clone();
    e3.external_tag = 3;
    let mut e4 = e1.clone();
    e4.external_tag = 4;

    let v8 = graph.add_node(e1.clone());
    let v9 = graph.add_node(e2.clone());
    let v10 = graph.add_node(e3.clone());
    let v11 = graph.add_node(e4.clone());

    let h = EdgeColor::from_particle(model.get_particle(&"H".to_string().into()));

    let b = EdgeColor::from_particle(model.get_particle(&"b".to_string().into()));
    let bbar = EdgeColor::from_particle(
        model
            .get_particle(&"b".to_string().into())
            .get_anti_particle(&model),
    );
    let g = EdgeColor::from_particle(model.get_particle(&"g".to_string().into()));
    let a = EdgeColor::from_particle(model.get_particle(&"a".to_string().into()));
    let eplus = EdgeColor::from_particle(model.get_particle(&"e+".to_string().into()));
    let eminus = EdgeColor::from_particle(model.get_particle(&"e-".to_string().into()));
    graph.add_edge(v0, v1, true, h).unwrap();
    graph.add_edge(v0, v5, true, b).unwrap();
    graph.add_edge(v1, v3, true, b).unwrap();
    graph.add_edge(v2, v4, true, g).unwrap();
    graph.add_edge(v2, v4, true, b).unwrap();
    graph.add_edge(v3, v2, true, b).unwrap();
    graph.add_edge(v3, v7, true, a).unwrap();
    graph.add_edge(v4, v1, true, b).unwrap();
    graph.add_edge(v5, v0, true, b).unwrap();
    graph.add_edge(v5, v6, true, a).unwrap();
    graph.add_edge(v6, v10, false, eplus).unwrap();
    graph.add_edge(v6, v11, false, eminus).unwrap();
    graph.add_edge(v8, v7, false, eplus).unwrap();
    graph.add_edge(v9, v7, false, eminus).unwrap();

    let (n_unresolved, unresolved_type) = filters.unresolved_cut_content(&model);
    assert!(!filters.contains_cut(&model, &graph, n_unresolved, &unresolved_type));

    let mut double_double_triangle = Graph::new();
    let v0 = double_double_triangle.add_node(e1.clone());
    let v1 = double_double_triangle.add_node(e2.clone());
    let v2 = double_double_triangle.add_node(e3.clone());
    let v3 = double_double_triangle.add_node(e4.clone());

    let v4 = double_double_triangle.add_node(bba.clone());
    let v5 = double_double_triangle.add_node(bba.clone());
    let v6 = double_double_triangle.add_node(bbH.clone());
    let v7 = double_double_triangle.add_node(bbH.clone());
    let v8 = double_double_triangle.add_node(bbg.clone());
    let v9 = double_double_triangle.add_node(bbg.clone());
    let v10 = double_double_triangle.add_node(bbg.clone());
    let v11 = double_double_triangle.add_node(bbg.clone());
    let v12 = double_double_triangle.add_node(epema.clone());
    let v13 = double_double_triangle.add_node(epema.clone());

    double_double_triangle
        .add_edge(v0, v13, true, eplus)
        .unwrap();
    double_double_triangle
        .add_edge(v1, v13, true, eminus)
        .unwrap();
    double_double_triangle.add_edge(v4, v13, false, a).unwrap();
    double_double_triangle
        .add_edge(v4, v11, true, bbar)
        .unwrap();
    double_double_triangle.add_edge(v4, v10, true, b).unwrap();
    double_double_triangle.add_edge(v10, v11, false, g).unwrap();
    double_double_triangle.add_edge(v6, v11, true, b).unwrap();
    double_double_triangle
        .add_edge(v6, v10, true, bbar)
        .unwrap();
    double_double_triangle.add_edge(v6, v7, true, h).unwrap();
    double_double_triangle.add_edge(v7, v8, true, bbar).unwrap();
    double_double_triangle.add_edge(v7, v9, true, b).unwrap();
    double_double_triangle.add_edge(v5, v8, true, b).unwrap();
    double_double_triangle.add_edge(v5, v9, true, bbar).unwrap();
    double_double_triangle.add_edge(v8, v9, true, g).unwrap();
    double_double_triangle.add_edge(v5, v12, true, a).unwrap();
    double_double_triangle
        .add_edge(v12, v2, true, eminus)
        .unwrap();
    double_double_triangle
        .add_edge(v12, v3, true, eplus)
        .unwrap();

    let mut coupling = HashMap::new();
    coupling.insert("QED".into(), 6);
    let mut pert = HashMap::new();
    pert.insert("QCD".into(), 2);
    let filters = FeynGen::new(FeynGenOptions {
        generation_type: GenerationType::CrossSection,
        initial_pdgs: vec![-11, 11],
        final_pdgs: vec![5, -5, 25],
        loop_count_range: (4, 4),
        symmetrize_initial_states: true,
        symmetrize_final_states: true,
        symmetrize_left_right_states: true,
        allow_symmetrization_of_external_fermions_in_amplitudes: false,
        max_multiplicity_for_fast_cut_filter: 6,
        amplitude_filters: FeynGenFilters(vec![
            FeynGenFilter::SelfEnergyFilter(Default::default()),
            FeynGenFilter::ParticleVeto(vec![
                23, 24, 9000001, 9000002, 9000003, 9000004, 12, 14, 16, 2, 4, 6, 3, 250, 251, 13,
                15,
            ]),
            FeynGenFilter::TadpolesFilter(Default::default()),
            FeynGenFilter::ZeroSnailsFilter(Default::default()),
        ]),
        cross_section_filters: FeynGenFilters(vec![
            FeynGenFilter::SelfEnergyFilter(Default::default()),
            FeynGenFilter::ParticleVeto(vec![
                23, 24, 9000001, 9000002, 9000003, 9000004, 12, 14, 16, 2, 4, 6, 3, 250, 251, 13,
                15,
            ]),
            FeynGenFilter::TadpolesFilter(Default::default()),
            FeynGenFilter::ZeroSnailsFilter(Default::default()),
            FeynGenFilter::PerturbativeOrders(pert),
            FeynGenFilter::CouplingOrders(coupling),
            FeynGenFilter::LoopCountRange((4, 4)),
        ]),
    });

    let (n_unresolved, unresolved_type) = filters.unresolved_cut_content(&model);
    assert!(!filters.contains_cut(
        &model,
        &double_double_triangle,
        n_unresolved,
        &unresolved_type
    ));
}
