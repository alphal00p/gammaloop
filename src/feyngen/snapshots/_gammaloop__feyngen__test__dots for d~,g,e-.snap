---
source: src/feyngen/test.rs
expression: dots
---
digraph G {
label="DIS0";
noverlap="scale"; layout="neato";
graph [ fontsize=10 ];
node [ fontsize=7,shape=circle,margin=0,height=0.01 ];
edge [ fontsize=5,arrowsize=0.4 ];
"v2" -> "v9" [label="p1 | d~ | Q(0) "];
"v1" -> "v10" [label="p2 | g | Q(1) "];
"v0" -> "v11" [label="p3 | e- | Q(2) "];
"v6" -> "v5" [label="p4 | d~ | Q(3) "];
"v7" -> "v4" [label="p5 | g | Q(4) "];
"v8" -> "v3" [label="p6 | e- | Q(5) "];
"v7" -> "v6" [label="q1 | d~ | Q(6) "];
"v6" -> "v8" [label="q2 | a | Q(7) "];
"v10" -> "v7" [label="q3 | d~ | Q(8) "];
"v11" -> "v8" [label="q4 | e- | Q(9) "];
"v9" -> "v10" [label="q5 | d~ | Q(10) "];
"v9" -> "v11" [label="q6 | a | Q(11) "];
}

digraph G {
label="DIS1";
noverlap="scale"; layout="neato";
graph [ fontsize=10 ];
node [ fontsize=7,shape=circle,margin=0,height=0.01 ];
edge [ fontsize=5,arrowsize=0.4 ];
"v2" -> "v9" [label="p1 | d~ | Q(0) "];
"v1" -> "v10" [label="p2 | g | Q(1) "];
"v0" -> "v11" [label="p3 | e- | Q(2) "];
"v6" -> "v5" [label="p4 | d~ | Q(3) "];
"v7" -> "v4" [label="p5 | g | Q(4) "];
"v8" -> "v3" [label="p6 | e- | Q(5) "];
"v6" -> "v8" [label="q1 | a | Q(6) "];
"v10" -> "v6" [label="q2 | d~ | Q(7) "];
"v9" -> "v7" [label="q3 | d~ | Q(8) "];
"v7" -> "v10" [label="q4 | d~ | Q(9) "];
"v11" -> "v8" [label="q5 | e- | Q(10) "];
"v9" -> "v11" [label="q6 | a | Q(11) "];
}

digraph G {
label="DIS2";
noverlap="scale"; layout="neato";
graph [ fontsize=10 ];
node [ fontsize=7,shape=circle,margin=0,height=0.01 ];
edge [ fontsize=5,arrowsize=0.4 ];
"v2" -> "v10" [label="p1 | d~ | Q(0) "];
"v1" -> "v10" [label="p2 | g | Q(1) "];
"v0" -> "v11" [label="p3 | e- | Q(2) "];
"v7" -> "v5" [label="p4 | d~ | Q(3) "];
"v8" -> "v4" [label="p5 | g | Q(4) "];
"v9" -> "v3" [label="p6 | e- | Q(5) "];
"v6" -> "v7" [label="q1 | d~ | Q(6) "];
"v8" -> "v6" [label="q2 | d~ | Q(7) "];
"v6" -> "v11" [label="q3 | a | Q(8) "];
"v7" -> "v9" [label="q4 | a | Q(9) "];
"v10" -> "v8" [label="q5 | d~ | Q(10) "];
"v11" -> "v9" [label="q6 | e- | Q(11) "];
}

digraph G {
label="DIS3";
noverlap="scale"; layout="neato";
graph [ fontsize=10 ];
node [ fontsize=7,shape=circle,margin=0,height=0.01 ];
edge [ fontsize=5,arrowsize=0.4 ];
"v2" -> "v9" [label="p1 | d~ | Q(0) "];
"v1" -> "v10" [label="p2 | g | Q(1) "];
"v0" -> "v11" [label="p3 | e- | Q(2) "];
"v7" -> "v5" [label="p4 | d~ | Q(3) "];
"v9" -> "v4" [label="p5 | g | Q(4) "];
"v8" -> "v3" [label="p6 | e- | Q(5) "];
"v6" -> "v7" [label="q1 | d~ | Q(6) "];
"v10" -> "v6" [label="q2 | d~ | Q(7) "];
"v6" -> "v11" [label="q3 | a | Q(8) "];
"v7" -> "v8" [label="q4 | a | Q(9) "];
"v11" -> "v8" [label="q5 | e- | Q(10) "];
"v9" -> "v10" [label="q6 | d~ | Q(11) "];
}

digraph G {
label="DIS5";
noverlap="scale"; layout="neato";
graph [ fontsize=10 ];
node [ fontsize=7,shape=circle,margin=0,height=0.01 ];
edge [ fontsize=5,arrowsize=0.4 ];
"v2" -> "v10" [label="p1 | d~ | Q(0) "];
"v1" -> "v10" [label="p2 | g | Q(1) "];
"v0" -> "v11" [label="p3 | e- | Q(2) "];
"v7" -> "v5" [label="p4 | d~ | Q(3) "];
"v8" -> "v4" [label="p5 | g | Q(4) "];
"v9" -> "v3" [label="p6 | e- | Q(5) "];
"v6" -> "v8" [label="q1 | d~ | Q(6) "];
"v10" -> "v6" [label="q2 | d~ | Q(7) "];
"v6" -> "v11" [label="q3 | a | Q(8) "];
"v8" -> "v7" [label="q4 | d~ | Q(9) "];
"v7" -> "v9" [label="q5 | a | Q(10) "];
"v11" -> "v9" [label="q6 | e- | Q(11) "];
}

digraph G {
label="DIS6";
noverlap="scale"; layout="neato";
graph [ fontsize=10 ];
node [ fontsize=7,shape=circle,margin=0,height=0.01 ];
edge [ fontsize=5,arrowsize=0.4 ];
"v2" -> "v9" [label="p1 | d~ | Q(0) "];
"v1" -> "v10" [label="p2 | g | Q(1) "];
"v0" -> "v11" [label="p3 | e- | Q(2) "];
"v7" -> "v5" [label="p4 | d~ | Q(3) "];
"v9" -> "v4" [label="p5 | g | Q(4) "];
"v8" -> "v3" [label="p6 | e- | Q(5) "];
"v9" -> "v6" [label="q1 | d~ | Q(6) "];
"v6" -> "v10" [label="q2 | d~ | Q(7) "];
"v6" -> "v11" [label="q3 | a | Q(8) "];
"v7" -> "v8" [label="q4 | a | Q(9) "];
"v10" -> "v7" [label="q5 | d~ | Q(10) "];
"v11" -> "v8" [label="q6 | e- | Q(11) "];
}

digraph G {
label="DIS8";
noverlap="scale"; layout="neato";
graph [ fontsize=10 ];
node [ fontsize=7,shape=circle,margin=0,height=0.01 ];
edge [ fontsize=5,arrowsize=0.4 ];
"v2" -> "v10" [label="p1 | d~ | Q(0) "];
"v1" -> "v10" [label="p2 | g | Q(1) "];
"v0" -> "v11" [label="p3 | e- | Q(2) "];
"v8" -> "v5" [label="p4 | d~ | Q(3) "];
"v8" -> "v4" [label="p5 | g | Q(4) "];
"v9" -> "v3" [label="p6 | e- | Q(5) "];
"v6" -> "v7" [label="q1 | d~ | Q(6) "];
"v10" -> "v6" [label="q2 | d~ | Q(7) "];
"v6" -> "v11" [label="q3 | a | Q(8) "];
"v7" -> "v8" [label="q4 | d~ | Q(9) "];
"v7" -> "v9" [label="q5 | a | Q(10) "];
"v11" -> "v9" [label="q6 | e- | Q(11) "];
}

digraph G {
label="DIS9";
noverlap="scale"; layout="neato";
graph [ fontsize=10 ];
node [ fontsize=7,shape=circle,margin=0,height=0.01 ];
edge [ fontsize=5,arrowsize=0.4 ];
"v2" -> "v9" [label="p1 | d~ | Q(0) "];
"v1" -> "v10" [label="p2 | g | Q(1) "];
"v0" -> "v11" [label="p3 | e- | Q(2) "];
"v7" -> "v5" [label="p4 | d~ | Q(3) "];
"v7" -> "v4" [label="p5 | g | Q(4) "];
"v8" -> "v3" [label="p6 | e- | Q(5) "];
"v6" -> "v7" [label="q1 | d~ | Q(6) "];
"v6" -> "v8" [label="q2 | a | Q(7) "];
"v10" -> "v6" [label="q3 | d~ | Q(8) "];
"v11" -> "v8" [label="q4 | e- | Q(9) "];
"v9" -> "v10" [label="q5 | d~ | Q(10) "];
"v9" -> "v11" [label="q6 | a | Q(11) "];
}

digraph G {
label="DIS10";
noverlap="scale"; layout="neato";
graph [ fontsize=10 ];
node [ fontsize=7,shape=circle,margin=0,height=0.01 ];
edge [ fontsize=5,arrowsize=0.4 ];
"v2" -> "v9" [label="p1 | d~ | Q(0) "];
"v1" -> "v10" [label="p2 | g | Q(1) "];
"v0" -> "v11" [label="p3 | e- | Q(2) "];
"v7" -> "v5" [label="p4 | d~ | Q(3) "];
"v7" -> "v4" [label="p5 | g | Q(4) "];
"v8" -> "v3" [label="p6 | e- | Q(5) "];
"v6" -> "v8" [label="q1 | a | Q(6) "];
"v9" -> "v6" [label="q2 | d~ | Q(7) "];
"v6" -> "v10" [label="q3 | d~ | Q(8) "];
"v10" -> "v7" [label="q4 | d~ | Q(9) "];
"v11" -> "v8" [label="q5 | e- | Q(10) "];
"v9" -> "v11" [label="q6 | a | Q(11) "];
}

digraph G {
label="DIS13";
noverlap="scale"; layout="neato";
graph [ fontsize=10 ];
node [ fontsize=7,shape=circle,margin=0,height=0.01 ];
edge [ fontsize=5,arrowsize=0.4 ];
"v2" -> "v9" [label="p1 | d~ | Q(0) "];
"v1" -> "v10" [label="p2 | g | Q(1) "];
"v0" -> "v11" [label="p3 | e- | Q(2) "];
"v10" -> "v5" [label="p4 | d~ | Q(3) "];
"v7" -> "v4" [label="p5 | g | Q(4) "];
"v8" -> "v3" [label="p6 | e- | Q(5) "];
"v6" -> "v7" [label="q1 | d~ | Q(6) "];
"v6" -> "v8" [label="q2 | a | Q(7) "];
"v9" -> "v6" [label="q3 | d~ | Q(8) "];
"v7" -> "v10" [label="q4 | d~ | Q(9) "];
"v11" -> "v8" [label="q5 | e- | Q(10) "];
"v9" -> "v11" [label="q6 | a | Q(11) "];
}

digraph G {
label="DIS14";
noverlap="scale"; layout="neato";
graph [ fontsize=10 ];
node [ fontsize=7,shape=circle,margin=0,height=0.01 ];
edge [ fontsize=5,arrowsize=0.4 ];
"v2" -> "v9" [label="p1 | d~ | Q(0) "];
"v1" -> "v10" [label="p2 | g | Q(1) "];
"v0" -> "v11" [label="p3 | e- | Q(2) "];
"v10" -> "v5" [label="p4 | d~ | Q(3) "];
"v7" -> "v4" [label="p5 | g | Q(4) "];
"v8" -> "v3" [label="p6 | e- | Q(5) "];
"v7" -> "v6" [label="q1 | d~ | Q(6) "];
"v6" -> "v8" [label="q2 | a | Q(7) "];
"v6" -> "v10" [label="q3 | d~ | Q(8) "];
"v9" -> "v7" [label="q4 | d~ | Q(9) "];
"v11" -> "v8" [label="q5 | e- | Q(10) "];
"v9" -> "v11" [label="q6 | a | Q(11) "];
}
